pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -375,32 +375,30 @@ func (r *SubnetSetReconciler) deleteSubnets(nsxSubnets []*model.VpcSubnet, delet
 	}
 	var deleteErrs []error
 	for _, nsxSubnet := range nsxSubnets {
-		lock := r.SubnetService.LockSubnet(nsxSubnet.Path)
-		func() {
-			defer r.SubnetService.UnlockSubnet(nsxSubnet.Path, lock)
-
-			portNums := len(r.SubnetPortService.GetPortsOfSubnet(*nsxSubnet.Id))
-			if portNums > 0 {
-				hasStalePort = true
-				log.Info("Skipped deleting NSX Subnet due to stale ports", "nsxSubnet", *nsxSubnet.Id)
-				return
-			}
 
-			if deleteBindingMaps {
-				if err := r.BindingService.DeleteSubnetConnectionBindingMapsByParentSubnet(nsxSubnet); err != nil {
-					deleteErr := fmt.Errorf("failed to delete NSX SubnetConnectionBindingMaps connected to NSX Subnet/%s: %+v", *nsxSubnet.Id, err)
-					deleteErrs = append(deleteErrs, deleteErr)
-					log.Error(deleteErr, "Skipping to next Subnet")
-					return
-				}
-			}
+		if !r.SubnetPortService.IsEmptySubnet(*nsxSubnet.Path) {
+			hasStalePort = true
+			log.Info("Skipped deleting NSX Subnet due to stale ports", "nsxSubnet", *nsxSubnet.Id)
+			continue
+		}
 
-			if err := r.SubnetService.DeleteSubnet(*nsxSubnet); err != nil {
-				deleteErr := fmt.Errorf("failed to delete NSX Subnet/%s: %+v", *nsxSubnet.Id, err)
+		if deleteBindingMaps {
+			if err = r.BindingService.DeleteSubnetConnectionBindingMapsByParentSubnet(nsxSubnet); err != nil {
+				deleteErr := fmt.Errorf("failed to delete NSX SubnetConnectionBindingMaps connected to NSX Subnet/%s: %+v", *nsxSubnet.Id, err)
 				deleteErrs = append(deleteErrs, deleteErr)
 				log.Error(deleteErr, "Skipping to next Subnet")
+				continue
 			}
-		}()
+		}
+
+		if err := r.SubnetService.DeleteSubnet(*nsxSubnet); err != nil {
```

Comment:
Is it possible to have a race that when this routine is deleting the subnet, and pod_controller/subnetport_controller is using the subnet to allocate a new port?

Reply:
Updated as discussed, thanks